src/Symfony/Component/Messenger/Envelope.php

```diff
@@ -0,0 +1,75 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger;
+
+/**
+ * A message wrapped in an envelope with items (configurations, markers, ...).
+ *
+ * @author Maxime Steinhausser <maxime.steinhausser@gmail.com>
+ *
+ * @experimental in 4.1
+ */
+final class Envelope
+{
+    private $items = array();
+    private $message;
+
+    /**
+     * @param object                  $message
+     * @param EnvelopeItemInterface[] $items
+     */
+    public function __construct($message, array $items = array())
+    {
+        $this->message = $message;
+        foreach ($items as $item) {
+            $this->with($item);
+        }
+    }
+
+    /**
+     * Wrap a message into an envelope if not already wrapped.
+     *
+     * @param Envelope|object $message
+     */
+    public static function wrap($message): self
+    {
+        return $message instanceof self ? $message : new self($message);
+    }
+
+    public function with(EnvelopeItemInterface $item): self
```

Comment:
`add()`? Usually `with()` implies immutable

Reply:
True. But quoting @sroze : 

> I'd prefer the envelope to wrap a message, with some additional ~configuration~ items I think.

I prefer `with` despite this convention. Once again, no strong opinion. Just waiting for changing everything for the umpteenth time.

Reply:
As already mentioned earlier, I have the same reaction with `with`. It makes sense to clone the message in it IMHO so it’s immutable.

Reply:
Fair enough, `Envelope` is now immutable 👍